### PR TITLE
Adds `jest/no-if`

### PR DESCRIPTION
**Pull request by cartogram**
Closes https://github.com/Shopify/eslint-plugin-shopify/issues/227

### CODE REVIEW

File: docs/rules/jest/no-if.md [hunk 1]
```diff
@@ -0,0 +1,34 @@
+# Prevent if statements in tests. (jest/no-if)
+This rule prevents the use of if statements in tests. 
+
+## Rule Details
+
+If statments in tests is usually an indication that a test is attempting to cover too much. Each branch of code executing within an if statement will likely be better served by a test devoted to it.
```
Comment (lemonmade) [inline docs/rules/jest/no-if.md]: Not only covering too much, but also possibly not actually running the test you think it is

File: lib/rules/jest/no-if.js [hunk 2]
```diff
@@ -0,0 +1,93 @@
+const {docsUrl} = require('../../utilities');
+
+const TEST_FUNCTION_NAMES = [
```
Comment (lemonmade) [inline lib/rules/jest/no-if.js]: We don't have this list for any other test?

File: lib/rules/jest/no-if.js [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+  'it',
+  'xit',
+  'fit',
+  'test',
+  'xtest',
+  'describe',
```
Comment (lemonmade) [inline lib/rules/jest/no-if.js]: I am not sure you need to include `describe` blocks

File: lib/rules/jest/no-if.js [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+        }
+        inCallExpression = true;
+      },
+      IfStatement(node) {
+        if (!inCallExpression) {
+          return;
+        }
+
+        context.report({
+          messageId: 'noIf',
```
Comment (lemonmade) [inline lib/rules/jest/no-if.js:43]: TIL

File: lib/rules/jest/no-if.js [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+
+  create(context) {
+    let inCallExpression = false;
+
+    return {
+      CallExpression(node) {
+        if (notTestFunction(node) || hasEmptyBody(node)) {
+          return;
+        }
+        inCallExpression = true;
```
Comment (lemonmade) [inline lib/rules/jest/no-if.js]: The iffy thing with doing this in scope is that nesting can throw things off. Nesting might not be too much of an issue here but you still might want to go with the more formal mechanism for this (state)

File: lib/rules/jest/no-if.js [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+        if (!inCallExpression) {
+          return;
+        }
+
+        context.report({
+          messageId: 'noIf',
+          node,
+        });
+      },
+      'CallExpression:exit': () => {
```
Comment (lemonmade) [inline lib/rules/jest/no-if.js]: I think this probably fails with:

```ts
it('should fail', () => {
  anotherCallExpressionToThrowThingsOff();
  if (ohShoot) {

  }
});
```

**Review by lemonmade — CHANGES_REQUESTED** [review]

File: yarn.lock [hunk 7]
```diff
@@ -3320,10 +3320,10 @@ mem@^1.1.0:
   dependencies:
     mimic-fn "^1.0.0"
 
-merge@1.2.0:
-  version "1.2.0"
-  resolved "https://registry.yarnpkg.com/merge/-/merge-1.2.0.tgz#7531e39d4949c281a66b8c5a6e0265e8b05894da"
-  integrity sha1-dTHjnUlJwoGma4xabgJl6LBYlNo=
+merge@1.2.1:
```
Comment (lemonmade) [inline yarn.lock:3323]: Why this change?

File: lib/rules/jest/no-if.js (hunk 3 above)
Comment (cartogram) [reply lib/rules/jest/no-if.js]: I figured I'd leave it cause it's not like we want to allow:
```
describe('SomeThing', () => {
  if(!someFlag) {
     it('testing', () => {
     //...
     })
  }
})

```

File: lib/rules/jest/no-if.js (hunk 5 above)
Comment (cartogram) [reply lib/rules/jest/no-if.js]: Can you clarify what you mean when you say `more formal mechanism for this (state)`?

Fixing the nesting part should be easy if I just keep track of the depth instead of just a boolean for being in a call expression or not. Maybe that is what you meant by state?

Comment (lemonmade) [reply lib/rules/jest/no-if.js]: I meant having it on the `context.state` or whatever that field is called so that it is scoped to the part of the tree where you are actually inside the test block (structurally, I don't think it makes much of a difference, but that's what the feature is there for)

Comment (lemonmade) [reply lib/rules/jest/no-if.js]: Wait actually maybe I am getting mixed up with Babel plugins :P

Comment (cartogram) [reply lib/rules/jest/no-if.js]: lol you are. I only know because I did the same thing. 😛

File: lib/rules/jest/no-if.js (hunk 2 above)
Comment (cartogram) [reply lib/rules/jest/no-if.js]: moved to a require from `utilities/jest`

File: yarn.lock (hunk 7 above)
Comment (cartogram) [reply yarn.lock:3323]: From here I guess: https://github.com/Shopify/eslint-plugin-shopify/commit/d45d8d07dfabe83e80d48661b235e72c0e123e9e

**Review by lemonmade — APPROVED** [review]
👍
